### DISCUSSION TITLE
Element-wise multiplication (hadamard product) support

### DISCUSSION THREAD

**Discussion by rbcarlos** [Xilinx/finn#460] [General]
Is there any way to implement element-wise multiplication (hadamard product) of two tensors? 

I have tried to implement this by simply doing: 
```
t3 = t1 * t2
```

When I try to streamline it however FINN throws an error during the streamlining:
```
Initializer for mul weights is not set.
```

**Comment by maltanar** [comment]
I assume you mean the case where both tensors are dynamic (i.e. not a constant)? That isn't possible right now, but could be enabled by doing the following:

1. Add a new HLS layer to `finn-hlslib` for multiplying two streams similar to https://github.com/Xilinx/finn-hlslib/blob/master/streamtools.h#L650
2. Add `HLSCustomOp` wrapper in `finn` similar to https://github.com/Xilinx/finn/blob/main/src/finn/custom_op/fpgadataflow/addstreams_batch.py
3. Add a convert-to-HLS transformation in `finn` similar to https://github.com/Xilinx/finn/blob/main/src/finn/transformation/fpgadataflow/convert_to_hls_layers.py#L1088 and a test https://github.com/Xilinx/finn/blob/main/tests/fpgadataflow/test_fpgadataflow_addstreams.py

If you meant the case where one tensor is dynamic and the other is a constant, it should be possible to implement this with existing functionality but might need some massaging (main suspect would be that the constant tensor is the 2nd op and not the 1st)
